server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RecipientRewriteTableProcessor.java

```diff
@@ -0,0 +1,309 @@
+
+    private ImmutableList<MailAddress> filterRecipients(ImmutableList<MappingData> mappingDatas, Predicate<MappingData> filterCondition) {
+        return FluentIterable.from(mappingDatas)
+            .filter(filterCondition)
+            .transformAndConcat(mailAddressesFromMappingData)
+            .toList();
+    }
+
+    private ImmutableList<MappingData> convertToMappingDatas(Mail mail, MailetContext mailetContext) {
+        return FluentIterable.from(mail.getRecipients())
+            .transform(convertToMappingData(mail, mailetContext))
+            .toList();
+    }
+
+    @VisibleForTesting
+    List<MailAddress> handleMappings(Mappings mappings, MailAddress sender, MailAddress recipient, MimeMessage message,
+                                                  MailetContext context, DomainList domainList) throws MessagingException {
+        ImmutableList<Mapping> addressMappingWithoutDomains = getAddressWithNoDomain(mappings, domainList);
+
+        ImmutableList<Mapping> newAddressMappings = convertToNewMappings(mappings, addressMappingWithoutDomains);
+
+        ImmutableList<MailAddress> mailAddresses = buildMailAddressFromMappingAddress(newAddressMappings);
+
+        forwardToRemoteAddress(sender, recipient, message, mailAddresses, context);
+
+        return getLocalAddresses(mailAddresses, context);
+    }
+
+    private ImmutableList<Mapping> convertToNewMappings(final Mappings mappings,
+            ImmutableList<Mapping> addressWithoutDomains) {
+        return FluentIterable.from(mappings)
+            .filter(haveDomain)
+            .append(addressWithoutDomains)
+            .toList();
+    }
+
+    private ImmutableList<MailAddress> getLocalAddresses(ImmutableList<MailAddress> mailAddresses,
+            final MailetContext context) {
+        return FluentIterable.from(mailAddresses)
+            .filter(isLocalServer(context))
+            .toList();
+    }
+
+    private ImmutableList<MailAddress> buildMailAddressFromMappingAddress(ImmutableList<Mapping> newMappings) {
+        return FluentIterable.from(newMappings)
+            .transform(mailAddressFromMapping)
+            .filter(mailAddressPresent)
+            .transform(mailAddressFromOptional)
+            .toList();
+    }
+
+    private ImmutableList<Mapping> getAddressWithNoDomain(Mappings mappings, DomainList domainList) throws MessagingException {
+        ImmutableList<Mapping> addressWithoutDomains = FluentIterable.from(mappings)
+            .filter(noneDomain)
+            .toList();
+        
+        if (!addressWithoutDomains.isEmpty()) {
+            final String defaultDomain = getDefaultDomain(domainList);
+
+            return FluentIterable.from(addressWithoutDomains)
+                .transform(appendDefaultDomain(defaultDomain))
+                .toList();
+        }
+        return ImmutableList.of();
+    }
+
+    private Function<MailAddress, MappingData> convertToMappingData(final Mail mail, final MailetContext mailetContext) {
```

Comment:
No. No Function in your return type.

Your return type should be List<MappingData>. Iterate over mail.getRecipients() to get it.